File: remove-all-adjacent-duplicates-in-string/Solution1047.java
// https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

import java.util.Stack;

class Solution1047 {

    // T: O(n), S: O(n)
    public String removeDuplicates(String str) {
        var stack = new Stack<Character>();
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);
            if (!stack.isEmpty() && stack.peek() == ch) {
                stack.pop();
            } else {
                stack.push(ch);
            }
        }
        var sb = new StringBuilder();
        for (var ch : stack) {
            sb.append(ch);
        }
        return sb.toString();
    }

    // T: O(n), S: O(n)
    public String removeDuplicates2(String s) {
        var st = new StringBuilder(); // Use sb as stack
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            int len = st.length();
            if (len > 0 && st.charAt(len - 1) == ch) {
                st.deleteCharAt(len - 1);
            } else {
                st.append(ch);
            }
        }
        return st.toString();
    }

    public static void main(String[] args) {
        var sol = new Solution1047();
        var s1 = "abbaca";
        var s2 = "azxxzy";
        System.out.println(sol.removeDuplicates(s1)); // "ca"
        System.out.println(sol.removeDuplicates2(s1)); // "ca"
        System.out.println(sol.removeDuplicates(s2)); // "ay"
        System.out.println(sol.removeDuplicates2(s2)); // "ay"
    }

}

File: remove-all-adjacent-duplicates-in-string/solution1047.py
# https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

class Solution1047:

    # T: O(n), S: O(n)
    def removeDuplicates(self, s: str) -> str:
        stack = []
        for ch in s:
            if stack and stack[-1] == ch:
                stack.pop()
            else:
                stack.append(ch)
        return ''.join(stack)

    # T: O(n), S: O(n)
    def removeDuplicates2(self, s: str) -> str:
        st = []  # Use list as stack
        for ch in s:
            if st and st[-1] == ch:
                st.pop()
            else:
                st.append(ch)
        return ''.join(st)


def main() -> None:
    sol = Solution1047()
    s1 = "abbaca"
    s2 = "azxxzy"
    print(sol.removeDuplicates(s1))  # "ca"
    print(sol.removeDuplicates2(s1))  # "ca"
    print(sol.removeDuplicates(s2))  # "ay"
    print(sol.removeDuplicates2(s2))  # "ay"


if __name__ == "__main__":
    main()
